### ISSUE TITLE
Should we start reading lines from `Terminal`?

### ISSUE THREAD

**Issue by jGleitz** [jGleitz/JUnit-KIT#15]
Referring to the [ILIAS question about error message lines](https://ilias.studium.kit.edu/ilias.php?ref_id=366572&cmdClass=ilobjforumgui&thr_pk=51401&cmd=viewThread&cmdNode=ed:5v&baseClass=ilRepositoryGUI), the Praktomat regards one call to `Terminal.printLine` as one line. 
Meanwhile, we handle one actual line printed to the console as one line when using `InteractiveConsoleTest`
## The problem

Say we would run the following session on my implementation of Sheet 6 Task C with the input file given on the task sheet on a tolerance of 0.3:

```
search OR(creator=galileocomputing, creator=ralf_reussner)
search isbn=1234
search title=java_ist_auch_eine_insel
```

We would probably assert the result like this:

```
expectedResultMatchers = getMatchers(
    is("creator=galileocomputing,title=java_ist_auch_eine_insel,year=unknown,true"),
    is("creator=unknown,title=grundkurs_programmieren_in_java,year=2007,false"),
    is("creator=ralf_reussner,title=unknown,year=2006,true"),

    startsWith("Error,"),

    is("creator=galileocomputing,title=java_ist_auch_eine_insel,year=unknown,true"),
    is("creator=unknown,title=grundkurs_programmieren_in_java,year=2007,false"),
    is("creator=ralf_reussner,title=unknown,year=2006,false")
);
multiLineTest(commands, expectedResultMatchers, "0.3", getFile(file));
```

Meanwhile, the output of my implementation would be:

```
creator=galileocomputing,title=java_ist_auch_eine_insel,year=unknown,true
creator=unknown,title=grundkurs_programmieren_in_java,year=2007,false
creator=ralf_reussner,title=unknown,year=2006,true
Error, expected a token matching 'creator|title|year' but found 'isbn' at character 4:
isbn=1234
   ^
creator=galileocomputing,title=java_ist_auch_eine_insel,year=unknown,true
creator=unknown,title=grundkurs_programmieren_in_java,year=2007,false
creator=ralf_reussner,title=unknown,year=2006,false
```

and therefore **fail the test while producing 100% correct output**.
## Solutions

I see two possible solutions for this:
### Only test errors at the end of a test

This would obviously fix the problem. To be able to do this, I already stopped to make sure that the tested class' output lines matches the amount of expected matchers. In the moment, the test is only failes if the tested class produces _less_ lines than expected, not if it produces _more_ lines. 
Take this example session:

```
search x
```

and this test:

```
multiLineTest(commands, startsWith("Error,"), "0.3", getFile(file));
```

If an implementation would call `Terminal.printLine()` _multiple times_, producing the following:

```
Error, bad search term
creator=galileocomputing,title=java_ist_auch_eine_insel,year=unknown,false
creator=unknown,title=grundkurs_programmieren_in_java,year=2007,false
creator=ralf_reussner,title=unknown,year=2006,false
```

the implementation would **pass the test while producing wrong output**.
So this solution feels more like a bad hack.
### Consider one call to `Terminal.printLine()` as one output line

Using Java reflection, we could take the program's output directly form the `Terminal` class. This would make sure, we test the way the Praktomat seems to do it. 
On the other hand, if somebody constructs the output lines by hand (divided by `\n`) and calls `Terminal.printLine()` _only once_, error messages would start getting bad readable. In this case, the test would output something like this:

```
java.lang.AssertionError: We ran a session on your interactive console
 that has been called with the command line arguments [0.5, textFile]
 with the following input file:

creator=galileocomputing,title=java_ist_auch_eine_insel
title=grundkursprogrammieren_in_java,year=2007
creator=ralf_reussner,year=2006

running the commands 

search creator=ralf_reussner
quit


 but got unexpected output:
Your program's output had not enough lines. Your output had 1 lines, expected were 3.
Expected was:

is "creator=galileocomputing,title=java_ist_auch_eine_insel,year=unknown,false"
is "creator=unknown,title=grundkursprogrammieren_in_java,year=2007,false"
is "creator=ralf_reussner,title=unknown,year=2006,true"

Your output was:

creator=galileocomputing,title=java_ist_auch_eine_insel,year=unknown,false
creator=unknown,title=grundkursprogrammieren_in_java,year=2007,false
creator=ralf_reussner,title=unknown,year=2006,true
```

I would regard such an error message as **higly confusing**.
## Your thoughts

I would highly appreciate hearing your thoughts on this matter!

**Comment by MartinLoeper** [comment]
Isn't there a way to ignore the new line characters? The praktomat does so, too, doesn't it?
Question:
Is "search search title=java_ist_auch_eine_insel" a typo? It does not look like a valid command.

**Comment by jGleitz** [comment]
"search search title=java_ist_auch_eine_insel" is a typo. Corrected it in the original question.

**Comment by jGleitz** [comment]
@MartinLoeper: In the moment, we take what the program prints to the console, split it at `System.getLineSeparator()` and the compare line by line.
I think what you are suggesting is my second option: Regard every call to `Terminal.printLine()` as one line and ignore `System.getLineSeparator()` in it.

**Comment by jGleitz** [comment]
As I realised writing #23, we always have to put an error test at the end of an test, as we cannot assume that the program continues to work after an error. Therefore, the above example is not valid. 
Nevertheless, it would be good to assure the tested class does not produce too much output.

**Comment by RomanLangrehr** [comment]
Well, i would appreciate when our tests work exactly the same way the Praktomat does, because if really somebody constructs the output lines by hand (divided by \n) and calls Terminal.printLine() only once, his solution may be correct, but as long as the Praktomat doesn't accept it, he will have to chnage it anyway.

**Comment by jGleitz** [comment]
@RomanLan didn't you succeed in an older task sheet by doing exactly that?

**Comment by jGleitz** [comment]
I implemented this behaviour in #32 (76f4cbc7a9fac82c6795f5263585b60d77d3a6aa). As it contains a major rewrite and known issues, please test it!
